src/Core/Grid/Definition/Factory/MerchandiseReturnProductsGridDefinitionFactory.php

```diff
@@ -0,0 +1,244 @@
+            ->add(
+                (new BulkActionColumn('merchandise_return_bulk'))
+                    ->setOptions([
+                        'bulk_field' => 'id_order_detail',
+                        'extra_fields' => [
+                            'id_customization',
+                        ],
+                    ])
+            )
+            ->add(
+                (new DataColumn('product_reference'))
+                    ->setName($this->trans('Reference', [], 'Admin.Global'))
+                    ->setOptions([
+                        'field' => 'product_reference',
+                    ])
+            )
+            ->add(
+                (new DataColumn('product_name'))
+                    ->setName($this->trans('Product name', [], 'Admin.Global'))
+                    ->setOptions([
+                        'field' => 'product_name',
+                    ])
+            )
+            ->add(
+                (new DataColumn('quantity'))
+                    ->setName($this->trans('Quantity', [], 'Admin.Global'))
+                    ->setOptions([
+                        'field' => 'product_quantity',
+                    ])
+            )
+            ->add(
+                (new MerchandiseReturnCustomizationColumn('customizations'))
+                    ->setName($this->trans('Customizations', [], 'Admin.Global'))
+                    ->setOptions([
+                        'customizations' => 'customizations',
+                    ])
+            )
+            ->add(
+                (new ActionColumn('actions'))
+                    ->setName($this->trans('Actions', [], 'Admin.Global'))
+                    ->setOptions([
+                        'actions' => (new RowActionCollection())
+                            ->add((new SubmitRowAction('delete'))
+                            ->setName($this->trans('Delete', [], 'Admin.Actions'))
+                            ->setIcon('delete')
+                            ->setOptions([
+                                'route' => 'admin_merchandise_returns_delete_product',
+                                'route_param_name' => 'merchandiseReturnId',
+                                'route_param_field' => 'id_order_return',
+                                'extra_route_params' => [
+                                    'merchandiseReturnDetailId' => 'id_order_detail',
+                                    'customizationId' => 'id_customization',
+                                ],
+                                'confirm_message' => $this->trans('Are you sure you want to delete the selected item?', [], 'Admin.Notifications.Warning'),
+                            ])
+                            ),
+                    ])
+            )
+        ;
+
+        return $columns;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getFilters()
+    {
+        /* @todo need to do something if merchandiseReturnId not found. Redirect to index? */
```

Comment:
You mean if someone loads page, which loads the erchandiseReturnProductsGridDefinitionFactory and no `merchandiseReturnId` is fetchable ?

Well ... if it makes no sense to have such a usecase
- throw an Exception from the merchandiseReturnId
- catch the Exception at controller level and redirect to index with a flash error message

Reply:
In `CmsPageCategoryDefinitionFactory` we use the RequestStack to obtain the `cmsCategoryParentId` and we put it as property of the class. Can we do the same here ?

This brings 3 benefits:
- it's possible to use this class in API or CLI context by providing a Mock RequestStack into the constructor; and you know that when `__construct()` is finished you don't need the dirty object anymore
- when `__construct()` is finished we don't need the RequestStack anymore and the RequestStack is stateful (it can evolve during code calls) so at least we won't have crazy bugs like "at X, my ID was 4 but 2 seconds later my ID is 7"
- you know in the `__construct()` whether or not you have the ID, so you can throw the exception at construct time instead of runtime

Reply:
Once again you might forget this `todo` if you resolve the comment before fixing it :stuck_out_tongue_closed_eyes: